File: 1371-MinimumRemoveToMakeValidParentheses/1371-MinimumRemoveToMakeValidParentheses.java
// Last updated: 3/26/2026, 1:49:52 AM
class Solution {
    public String minRemoveToMakeValid(String s) {
        LinkedList<Integer> openIndex = new LinkedList<>();
        LinkedList<Integer> closeIndex = new LinkedList<>();
        int openCount = 0, closeCount = 0;
        
        for(int i=0;i<s.length();i++){
            if(s.charAt(i)=='(')
            {
                openIndex.add(i); 
            }
            else if(s.charAt(i)==')'){
                if(openIndex.size()>0){
                    openIndex.remove();
                }
                else{
                    closeIndex.add(i);
                } 
            } 
        }
        StringBuilder sb = new StringBuilder();
        for(int i=0;i<s.length();i++){
            char c = s.charAt(i);
            if(c== '(' && openIndex.contains(i)){
                continue;
            }
            else if(c==')' && closeIndex.contains(i))
                continue;
            else
                sb.append(c);
        }
        return sb.toString();
    }
}

File: 1371-MinimumRemoveToMakeValidParentheses/1371-MinimumRemoveToMakeValidParentheses.cs
// Last updated: 3/26/2026, 1:49:50 AM
public class Solution {
    public string MinRemoveToMakeValid(string s) {
        int open = 0;
        int balance = 0;
        StringBuilder sb = new StringBuilder();

        for(int i=0;i<s.Length;i++)
        {
            if(s[i]==')')
            {
                if(balance>0)
                {
                    balance--;
                }
                else
                {
                    continue;
                }
            }
            else
            if(s[i]=='(')
            {
                open++;
                balance++;
            }
            sb.Append(s[i]);
        }
        StringBuilder result = new StringBuilder();
        int keep = open - balance;
        foreach(char c in sb.ToString())
        {
            if(c=='(')
            {
                keep--;
                if(keep<0)
                    continue;
            }
            result.Append(c);
        }

        return result.ToString();
    }
}